README.md
# measure [![Build Status](https://travis-ci.org/dubbs/measure.png)](https://travis-ci.org/dubbs/measure)

measure is a library for converting between units and unit systems.

It was created out of the need to adjust recipe amounts based on *servings* on the client-side, while also providing various output types.

There are some inconsistencies between US and Imperial systems.  One example would be, a UK "pint" is 568ml, while a US "pint" is 473ml.  There are many more.  Because of this reason, it is important to state the unit system your are starting with when developing a recipe, unless you are using metric.

## Unit Systems

### Customary (US)

To use US customary units (default):

```js
measure('1 teaspoon').milliliters() // 4.93
measure('1 teaspoon', 'US').milliliters() // 4.93
```

### Imperial (UK and CAN)

To use the Imperial units:

```js
measure('1 teaspoon', 'Imperial').milliliters() // 4.74
```

### Volume

```js
// US and Imperial

// long inputs
measure('1 teaspoon').teaspoons() // 1
measure('1 tablespoon').teaspoons() // 3

// abbrev. inputs
measure('1 1/2 tsp.').teaspoons() // 1.5
measure('1 tbsp.').teaspoons() // 3

// multiple inputs
measure('2 cups and 1 pint').quarts() // 1

// Metric

// long inputs
measure('1 milliliter').milliliters() // 1
measure('1 liter').milliliters() // 1000

// abbrev. inputs
measure('1.5 ml').milliliters() // 1.5
measure('1 l').milliliters() // 1000

// multiple inputs
measure('1 liter and 1 centiliter').milliliters() // 1010
```

### Mass

```js
// US and Imperial

// long inputs
measure('1 ounce').ounces() // 1
measure('1 pound').ounces() // 16

// abbrev. inputs
measure('1 1/2 oz.').ounces() // 1.5
measure('1 lbs.').pounds() // 1

// multiple inputs
measure('7 pounds 8 ounces').ounces() // 120

// Metric

// long inputs
measure('1 gram').grams() // 1
measure('1 kilogram').kilograms() // 1

// abbrev. inputs
measure('1 g.').grams() // 1
measure('1 kg.').kilograms() // 1

// multiple inputs
measure('3 kilograms and 2 grams').kilograms() // 3.002
```

## Operations

```js
measure('1 teaspoon').add('1 teaspoon').teaspoons(); // 2
measure('3 1/4 teaspoon').subtract('2 3/4 teaspoon').teaspoons(); // .5
measure('1 1/2 teaspoon').multiply(3).teaspoons(); // 4.5
measure('6 1/2 teaspoon').divide(2).teaspoons(); // 3.25
```

## Contributing

Grunt is used for build and tests, JSHint to enforce style.

### Install Dependencies

```
npm install && bower install
```

### Run Tests

```
grunt test
```

### Build

```
grunt build
```

## Influences

The great moment.js




changelog.md
<li> <a href="http://github.com/dubbs/measure/commit/4e9a170364041e1f9c492f4f4f518852920dcd89">view commit &bull;</a> add density spec and conversion helpers, #changelog</li> 


src/measure.js
(function (window) {

  // US Customary
  function roundUnits(ratio) {
    return Math.round((ratio + 0.00001) * 100) / 100;
  }

  var mlPerGallon = 3785.41;
  var mlPerOunce = 29.5735;

  var volume = {
    teaspoons: roundUnits(mlPerOunce / 6),
    tablespoons: roundUnits(mlPerOunce / 2),
    fluidounces: roundUnits(mlPerOunce),
    shots: roundUnits(mlPerOunce * 1.5),
    gills: roundUnits(mlPerOunce * 4),
    cups: roundUnits(mlPerGallon / 16),
    pints: roundUnits(mlPerGallon / 8),
    fifths: roundUnits(mlPerGallon / 5),
    quarts: roundUnits(mlPerGallon / 4),
    gallons: roundUnits(mlPerGallon)
  };

  // US Avoirdupois

  var gPerPound = 453.592;

  var mass = {
    grains: roundUnits(gPerPound / 7000),
    drams: roundUnits(gPerPound / 256),
    ounces: roundUnits(gPerPound / 16),
    pounds: roundUnits(gPerPound),
    quarters: roundUnits(gPerPound * 25),
    hundredweights: roundUnits(gPerPound * 100),
    ton: roundUnits(gPerPound * 2000)
  };

  window.MeasureVolumeUSCustomary = volume;
  window.MeasureMassUSCustomary = mass;

}(this));


(function (window) {

  // Imperial

  function roundUnits(ratio) {
    return Math.round((ratio + 0.00001) * 100) / 100;
  }

  var mlPerGallon = 4546.09; // 10 pounds of water in 1824
  var mlPerOunce = 28.4131;

  var volumes = {
    teaspoons: roundUnits(mlPerOunce / 6),
    tablespoons: roundUnits(mlPerOunce / 2),
    fluidounces: roundUnits(mlPerOunce),
    shots: roundUnits(mlPerOunce * 1.5),
    gills: roundUnits(mlPerGallon / 32),
    cups: roundUnits(mlPerGallon / 16),
    pints: roundUnits(mlPerGallon / 8),
    fifths: roundUnits(mlPerGallon / 5),
    quarts: roundUnits(mlPerGallon / 4),
    gallons: roundUnits(mlPerGallon)
  };

  // Imperial Avoirdupois 

  var gPerPound = 453.592;

  var mass = {
    drams: roundUnits(gPerPound / 256),
    ounces: roundUnits(gPerPound / 16),
    pounds: roundUnits(gPerPound),
    stones: roundUnits(gPerPound * 14),
    quarters: roundUnits(gPerPound * 28),
    hundredweights: roundUnits(gPerPound * 112),
    ton: roundUnits(gPerPound * 2240)
  };

  window.MeasureVolumeImperial = volumes;
  window.MeasureMassImperial = mass;

}(this));

(function (window) {

  function Measure(options) {
    this.ml = options && options.ml || 0;
    this.g = options && options.g || 0;
    this.d = options && options.d || 1;
  }

  // density
  Measure.prototype.density = function () {
    return this.d;
  };

  // helpers
  Measure.prototype.toVolume = function () {
    if (this.g) {
      this.ml = this.g / this.d;
      this.g = 0;
    }
    return this;
  };
  Measure.prototype.toMass = function () {
    if (this.ml) {
      this.g = this.d * this.ml;
      this.ml = 0;
    }
    return this;
  };

  // volume
  Measure.prototype.milliliters = function () {
    return this.ml;
  };
  Measure.prototype.centiliters = function () {
    return this.ml / 10;
  };
  Measure.prototype.deciliters = function () {
    return this.ml / 100;
  };
  Measure.prototype.liters = function () {
    return this.ml / 1000;
  };
  Measure.prototype.decaliters = function () {
    return this.ml / 10000;
  };
  Measure.prototype.hectoliters = function () {
    return this.ml / 100000;
  };
  Measure.prototype.kiloliters = function () {
    return this.ml / 1000000;
  };
  Measure.prototype.drops = function () {
    return this.totalByUnit('drops');
  };
  Measure.prototype.teaspoons = function () {
    return this.totalByUnit('teaspoons');
  };
  Measure.prototype.tablespoons = function () {
    return this.totalByUnit('tablespoons');
  };
  Measure.prototype.fluidounces = function () {
    return this.totalByUnit('fluidounces');
  };
  Measure.prototype.jiggers = function () {
    return this.totalByUnit('jiggers');
  };
  Measure.prototype.gills = function () {
    return this.totalByUnit('gills');
  };
  Measure.prototype.cups = function () {
    return this.totalByUnit('cups');
  };
  Measure.prototype.pints = function () {
    return this.totalByUnit('pints');
  };
  Measure.prototype.fifths = function () {
    return this.totalByUnit('fifths');
  };
  Measure.prototype.quarts = function () {
    return this.totalByUnit('quarts');
  };
  Measure.prototype.gallons = function () {
    return this.totalByUnit('gallons');
  };

	// mass
  Measure.prototype.grams = function () {
    return this.g;
  };
  Measure.prototype.kilograms = function () {
    return this.g / 1000;
  };
  Measure.prototype.drams = function () {
    return this.totalMassByUnit('drams');
  };
  Measure.prototype.ounces = function () {
    return this.totalMassByUnit('ounces');
  };
  Measure.prototype.pounds = function () {
    return this.totalMassByUnit('pounds');
  };
  Measure.prototype.quarters = function () {
    return this.totalMassByUnit('quarters');
  };

	// count
  Measure.prototype.units = function () {
    return this.totalByUnit('units');
  };

	// operations
  Measure.prototype.add = function (input) {
    var options = Measure.parseOptionsFromString(input);
    if (options.ml) {
      this.ml += options.ml;
    }
    if (options.g) {
      this.g += options.g;
    }
    return this;
  };
  Measure.prototype.subtract = function (input) {
    var options = Measure.parseOptionsFromString(input);
    if (options.ml) {
      this.ml -= options.ml;
    }
    if (options.g) {
      this.g -= options.g;
    }
    return this;
  };
  Measure.prototype.multiply = function (input) {
    this.ml *= input;
    this.g *= input;
    return this;
  };
  Measure.prototype.divide = function (input) {
    this.ml /= input;
    this.g /= input;
    return this;
  };

  // totals
  Measure.prototype.totalByUnit = function (unit) {
    var ratio = this.ml / Measure.volume[unit];
    return Math.round((ratio + 0.00001) * 100) / 100;
  };
  Measure.prototype.totalMassByUnit = function (unit) {
    var ratio = this.g / Measure.mass[unit];
    return Math.round((ratio + 0.00001) * 100) / 100;
  };

	// lexer
  Measure.parseOptionsFromString = function (input) {

    var lexer = new window.Lexer();

    var num = 0;
    var obj = {ml: 0, g: 0};

    var volume = this.volume;
    var mass = this.mass;

    lexer.addRule(/[0-9.\/ -]+/g, function (lexeme) {
      // add mixed numbers
      lexeme = lexeme.trim().replace('-', ' ').split(' ').join('+');
      num = eval(lexeme);
    });
    // volume
    // - metric
    lexer.addRule(/millilit(re|er)s?|ml/g, function () {
      obj.ml += num;
    });
    lexer.addRule(/centilit(re|er)s?|cl/g, function () {
      obj.ml += num * 10;
    });
    lexer.addRule(/decilit(re|er)s?|dl/g, function () {
      obj.ml += num * 100;
    });
    lexer.addRule(/lit(re|er)s?|l\s|l$/g, function () {
      obj.ml += num * 1000;
    });
    lexer.addRule(/decalit(re|er)s?|dal/g, function () {
      obj.ml += num * 10000;
    });
    lexer.addRule(/hectolit(re|er)s?|hl/g, function () {
      obj.ml += num * 100000;
    });
    lexer.addRule(/kilolit(re|er)s?|kl/g, function () {
      obj.ml += num * 1000000;
    });
    // - customary
    lexer.addRule(/(teaspoons?|tsp\.?|t\.)/g, function () {
      obj.ml += num * volume.teaspoons;
    });
    lexer.addRule(/(tablespoons?|tbsp\.?|T\.)/g, function () {
      obj.ml += num * volume.tablespoons;
    });
    lexer.addRule(/(fluidounces?|fl\.oz\.)/g, function () {
      obj.ml += num * volume.fluidounces;
    });
    lexer.addRule(/(shots?)/g, function () {
      obj.ml += num * volume.shots;
    });
    lexer.addRule(/(gills?|gi\.)/g, function () {
      obj.ml += num * volume.gills;
    });
    lexer.addRule(/(cups?|cup$|C)/g, function () {
      obj.ml += num * volume.cups;
    });
    lexer.addRule(/(pints?|pt\.)/g, function () {
      obj.ml += num * volume.pints;
    });
    lexer.addRule(/(fifths?)/g, function () {
      obj.ml += num * volume.fifths;
    });
    lexer.addRule(/(quarts?\s|quart$|qt\.)/g, function () {
      obj.ml += num * volume.quarts;
    });
    lexer.addRule(/(gallons?|gal\.)/g, function () {
      obj.ml += num * volume.gallons;
    });
    // mass
    // - metric
    lexer.addRule(/(kilograms?|kg\.?)/g, function () {
      obj.g += num * 1000;
    });
    lexer.addRule(/(grams?|g\.?)/g, function () {
      obj.g += num;
    });
    // - customary
    lexer.addRule(/(drams?|dr)/g, function () {
      obj.g += num * mass.drams;
    });
    lexer.addRule(/(ounces?|oz\.)/g, function () {
      obj.g += num * mass.ounces;
    });
    lexer.addRule(/(pounds?|lbs\.?)/g, function () {
      obj.g += num * mass.pounds;
    });
    lexer.addRule(/(quarters?|qr)/g, function () {
      obj.g += num * mass.quarters;
    });

    lexer.addRule(/[^0-9]+/g, function () {
      // if haven't set another value yet, assume its an arbitrary unit
      if (obj.ml === 0 && obj.g === 0) {
        obj.units = num;
      }
    });
    lexer.addRule(/\s/g, function () {});

    lexer.setInput(input);

    lexer.lex();

    // cleanup
    if (obj.ml === 0) {
      delete obj.ml;
    }
    if (obj.g === 0) {
      delete obj.g;
    }

    return obj;

  };

	// options
  Measure.setUnitSystem = function (unitSystem) {
    switch (unitSystem) {
    case 'Imperial':
      Measure.volume = window.MeasureVolumeImperial;
      Measure.mass = window.MeasureMassImperial;
      break;
    default: // US
      Measure.volume = window.MeasureVolumeUSCustomary;
      Measure.mass = window.MeasureMassUSCustomary;
      break;
    }
  };

	// create
  Measure.createFromString = function (options) {
    options = Measure.parseOptionsFromString(options);
    return new Measure(options);
  };
  Measure.createFromObject = function (options) {
    return new Measure(options);
  };

  var measure = function (options, unitSystem) {
    if (unitSystem) {
      Measure.setUnitSystem(unitSystem);
    } else {
      Measure.setUnitSystem('US');
    }
    if (typeof options === 'string') {
      return Measure.createFromString(options);
    }
    if (typeof options === 'object') {
      return Measure.createFromObject(options);
    }
    return new Measure();
  };

  window.measure = measure;

  window.Measure = Measure;

}(this));
